{
 "cells": [
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# DNA Sequence Classification\n",
    "\n",
    "DNA sequence prediction is a crucial task in bioinformatics, enabling researchers to analyze genetic patterns, predict mutations, and model gene structures. This dataset can be used to implement three machine learning approaches to predict nucleotide sequences: N-Gram, LSTM, and Transformer models.\n",
    "\n",
    "We use nucleotide sequences of human genes from the NCBI Gene Database. The dataset consists of:\n",
    "\n",
    "1. Gene symbols, descriptions, and types.\n",
    "2. Nucleotide sequences represented as A, T, C, G.\n",
    "3. Train-validation split: 80% training, 20% testing.\n",
    "\n",
    "Source: https://www.kaggle.com/datasets/harshvardhan21/dna-sequence-prediction"
   ],
   "id": "2ecd474989cfe337"
  },
  {
   "cell_type": "code",
   "id": "initial_id",
   "metadata": {
    "collapsed": true,
    "ExecuteTime": {
     "end_time": "2025-11-09T17:10:47.812440Z",
     "start_time": "2025-11-09T17:10:47.796779Z"
    }
   },
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ],
   "outputs": [],
   "execution_count": 1
  },
  {
   "cell_type": "code",
   "id": "906166dc",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-09T17:10:48.027385Z",
     "start_time": "2025-11-09T17:10:47.816752Z"
    }
   },
   "source": [
    "import os\n",
    "\n",
    "df_all = \"Data\"  # folder tempat file CSV kamu\n",
    "\n",
    "for name_file in os.listdir(df_all):\n",
    "    full_path = os.path.join(df_all, name_file)\n",
    "\n",
    "    # cek: beneran file + hanya CSV\n",
    "    if os.path.isfile(full_path) and name_file.endswith(\".csv\"):\n",
    "        print(full_path)  # path lengkap\n",
    "        print(f\"=== {name_file} ===\")  # nama file saja\n",
    "\n",
    "        df = pd.read_csv(full_path)\n",
    "        \n",
    "        print(f\"{name_file}: {df.shape}\")  # (baris, kolom)\n",
    "        print(\"-\" * 40)"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Data\\test.csv\n",
      "=== test.csv ===\n",
      "test.csv: (8326, 7)\n",
      "----------------------------------------\n",
      "Data\\train.csv\n",
      "=== train.csv ===\n",
      "train.csv: (22593, 7)\n",
      "----------------------------------------\n",
      "Data\\validation.csv\n",
      "=== validation.csv ===\n",
      "validation.csv: (4577, 7)\n",
      "----------------------------------------\n"
     ]
    }
   ],
   "execution_count": 2
  },
  {
   "cell_type": "markdown",
   "id": "1300e584",
   "metadata": {},
   "source": [
    "### Data Cleaning and Preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "id": "58656c8a",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-09T17:10:48.593283Z",
     "start_time": "2025-11-09T17:10:48.375830Z"
    }
   },
   "source": [
    "test_df = pd.read_csv(\"Data/test.csv\", encoding='ascii')\n",
    "train_df = pd.read_csv(\"Data/train.csv\", encoding='ascii')\n",
    "val_df = pd.read_csv(\"Data/validation.csv\", encoding='ascii')\n",
    "\n",
    "train_df.head()"
   ],
   "outputs": [
    {
     "data": {
      "text/plain": [
       "   Unnamed: 0  NCBIGeneID        Symbol  \\\n",
       "0           0   106481178      RNU4-21P   \n",
       "1           1   123477792  LOC123477792   \n",
       "2           2   113174975  LOC113174975   \n",
       "3           3   116216107  LOC116216107   \n",
       "4           4       28502      IGHD2-21   \n",
       "\n",
       "                                         Description           GeneType  \\\n",
       "0               RNA, U4 small nuclear 21, pseudogene             PSEUDO   \n",
       "1                Sharpr-MPRA regulatory region 12926  BIOLOGICAL_REGION   \n",
       "2                 Sharpr-MPRA regulatory region 7591  BIOLOGICAL_REGION   \n",
       "3  CRISPRi-validated cis-regulatory element chr10...  BIOLOGICAL_REGION   \n",
       "4                immunoglobulin heavy diversity 2-21              OTHER   \n",
       "\n",
       "  GeneGroupMethod                                 NucleotideSequence  \n",
       "0   NCBI Ortholog  <AGCTTAGCACAGTGGCAGTATCATAGGCAGTGAGGTTTATCCGAG...  \n",
       "1   NCBI Ortholog  <CTGGAGCGGCCACGATGTGAACTGTCACCGGCCACTGCTGCTCCG...  \n",
       "2   NCBI Ortholog  <TTCCCAATTTTTCCTCTGCTTTTTAATTTTCTAGTTTCCTTTTTC...  \n",
       "3   NCBI Ortholog  <CGCCCAGGCTGGAGTGCAGTGGCGCCATCTCGGCTCACTGCAGGC...  \n",
       "4   NCBI Ortholog                     <AGCATATTGTGGTGGTGACTGCTATTCC>  "
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>NCBIGeneID</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>Description</th>\n",
       "      <th>GeneType</th>\n",
       "      <th>GeneGroupMethod</th>\n",
       "      <th>NucleotideSequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>106481178</td>\n",
       "      <td>RNU4-21P</td>\n",
       "      <td>RNA, U4 small nuclear 21, pseudogene</td>\n",
       "      <td>PSEUDO</td>\n",
       "      <td>NCBI Ortholog</td>\n",
       "      <td>&lt;AGCTTAGCACAGTGGCAGTATCATAGGCAGTGAGGTTTATCCGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>123477792</td>\n",
       "      <td>LOC123477792</td>\n",
       "      <td>Sharpr-MPRA regulatory region 12926</td>\n",
       "      <td>BIOLOGICAL_REGION</td>\n",
       "      <td>NCBI Ortholog</td>\n",
       "      <td>&lt;CTGGAGCGGCCACGATGTGAACTGTCACCGGCCACTGCTGCTCCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>113174975</td>\n",
       "      <td>LOC113174975</td>\n",
       "      <td>Sharpr-MPRA regulatory region 7591</td>\n",
       "      <td>BIOLOGICAL_REGION</td>\n",
       "      <td>NCBI Ortholog</td>\n",
       "      <td>&lt;TTCCCAATTTTTCCTCTGCTTTTTAATTTTCTAGTTTCCTTTTTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>116216107</td>\n",
       "      <td>LOC116216107</td>\n",
       "      <td>CRISPRi-validated cis-regulatory element chr10...</td>\n",
       "      <td>BIOLOGICAL_REGION</td>\n",
       "      <td>NCBI Ortholog</td>\n",
       "      <td>&lt;CGCCCAGGCTGGAGTGCAGTGGCGCCATCTCGGCTCACTGCAGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>28502</td>\n",
       "      <td>IGHD2-21</td>\n",
       "      <td>immunoglobulin heavy diversity 2-21</td>\n",
       "      <td>OTHER</td>\n",
       "      <td>NCBI Ortholog</td>\n",
       "      <td>&lt;AGCATATTGTGGTGGTGACTGCTATTCC&gt;</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 3
  },
  {
   "cell_type": "code",
   "id": "be758efd",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-09T17:10:48.702117Z",
     "start_time": "2025-11-09T17:10:48.687533Z"
    }
   },
   "source": [
    "#train_df = train_df.drop(columns=[\"Unnamed: 0\"])\n",
    "#test_df = test_df.drop(columns=[\"Unnamed: 0\"])\n",
    "#val_df = val_df.drop(columns=[\"Unnamed: 0\"])"
   ],
   "outputs": [],
   "execution_count": 4
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-09T17:10:48.786333Z",
     "start_time": "2025-11-09T17:10:48.777504Z"
    }
   },
   "cell_type": "code",
   "source": "print(type(train_df))",
   "id": "2bcce31826603f1e",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n"
     ]
    }
   ],
   "execution_count": 5
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "Dataset yang didapat sudah bersih dan siap untuk ditraining. terdapat 3 file csv, yaitu\n",
    "- train.csv\n",
    "- test.csv\n",
    "- validation.csv"
   ],
   "id": "b3d7715bf3141701"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "### Classification Models",
   "id": "3511ac3cc3bcf8e5"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-09T17:10:49.313209Z",
     "start_time": "2025-11-09T17:10:48.851276Z"
    }
   },
   "cell_type": "code",
   "source": [
    "from collections import Counter\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.metrics import classification_report\n",
    "from itertools import product\n",
    "\n",
    "# fitur dan label\n",
    "X_train = train_df[\"NucleotideSequence\"].astype(str)\n",
    "y_train = train_df[\"GeneType\"]\n",
    "\n",
    "X_val = val_df[\"NucleotideSequence\"].astype(str)\n",
    "y_val = val_df[\"GeneType\"]\n",
    "\n",
    "X_test = test_df[\"NucleotideSequence\"].astype(str)"
   ],
   "id": "dcc1f119cbc02606",
   "outputs": [],
   "execution_count": 6
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "### Sequence + buat k-mer",
   "id": "fa2c5b5d4dd515bb"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-09T17:10:49.345721Z",
     "start_time": "2025-11-09T17:10:49.336336Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def clean_sequence(seq: str) -> str:\n",
    "    seq = str(seq).upper().strip()\n",
    "    seq = seq.replace(\"<\",\"\").replace(\">\", \"\")\n",
    "    return seq\n",
    "\n",
    "def get_kmers(seq, k=3):\n",
    "    seq = clean_sequence(seq)\n",
    "    if len(seq) < k:\n",
    "        return []\n",
    "    return [seq[i:i+k] for i in range(len(seq) - k + 1)]\n",
    "\n",
    "def kmer_analyzer(seq):\n",
    "    return get_kmers(seq, k=3)"
   ],
   "id": "e5552e49cd1710a0",
   "outputs": [],
   "execution_count": 7
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "#### Pipeline models classic (k-mer + TF-IDF + SVM)",
   "id": "fbc0b4aec76e8bfe"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-09T17:10:49.422116Z",
     "start_time": "2025-11-09T17:10:49.389993Z"
    }
   },
   "cell_type": "code",
   "source": [
    "from sklearn.pipeline import Pipeline\n",
    "from sklearn.feature_extraction.text import CountVectorizer, TfidfTransformer\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.metrics import accuracy_score, classification_report\n",
    "\n",
    "pipeline = Pipeline([\n",
    "    (\"cv\", CountVectorizer(\n",
    "        analyzer=kmer_analyzer,\n",
    "        max_features=5000,       # batasi jumlah fitur\n",
    "        min_df=5)),\n",
    "    (\"tfidf\", TfidfTransformer()),\n",
    "    (\"clf\", RandomForestClassifier(\n",
    "        n_estimators=200,        # jumlah pohon\n",
    "        max_depth=None,          # biarkan None dulu (boleh kamu batasi misal 20)\n",
    "        n_jobs=-1,               # pakai semua core CPU\n",
    "        random_state=42\n",
    "    ))\n",
    "])"
   ],
   "id": "343c674d08296e65",
   "outputs": [],
   "execution_count": 8
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "### Results",
   "id": "c0c0fe40f3e5a2dc"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-09T17:11:47.342981Z",
     "start_time": "2025-11-09T17:10:49.460282Z"
    }
   },
   "cell_type": "code",
   "source": [
    "pipeline.fit(X_train, y_train)\n",
    "\n",
    "y_pred = pipeline.predict(X_val)\n",
    "\n",
    "print(\"Validation accuracy:\", accuracy_score(y_val, y_pred))\n",
    "print(classification_report(y_val,y_pred))"
   ],
   "id": "d26c18a40d7c73e2",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validation accuracy: 0.9997815162770374\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "BIOLOGICAL_REGION       1.00      1.00      1.00      1398\n",
      "            OTHER       1.00      1.00      1.00        88\n",
      "   PROTEIN_CODING       1.00      1.00      1.00       101\n",
      "           PSEUDO       1.00      1.00      1.00      2133\n",
      "            ncRNA       1.00      1.00      1.00       516\n",
      "             rRNA       1.00      1.00      1.00         8\n",
      "            snRNA       0.96      1.00      0.98        22\n",
      "           snoRNA       1.00      1.00      1.00       239\n",
      "             tRNA       1.00      1.00      1.00        72\n",
      "\n",
      "         accuracy                           1.00      4577\n",
      "        macro avg       1.00      1.00      1.00      4577\n",
      "     weighted avg       1.00      1.00      1.00      4577\n",
      "\n"
     ]
    }
   ],
   "execution_count": 9
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-09T17:11:49.171819Z",
     "start_time": "2025-11-09T17:11:47.469989Z"
    }
   },
   "cell_type": "code",
   "source": [
    "import joblib  # biasanya sudah ada dari scikit-learn\n",
    "\n",
    "# misal kamu sudah punya:\n",
    "# pipeline.fit(X_train_full, y_train_full)\n",
    "\n",
    "# Simpan model ke file\n",
    "joblib.dump(pipeline, \"dna_classifier_pipeline.joblib\", compress=3)\n",
    "\n",
    "print(\"Model saved as dna_classifier_pipeline.joblib\")"
   ],
   "id": "1feb3882c9bf41b3",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Model saved as dna_classifier_pipeline.joblib\n"
     ]
    }
   ],
   "execution_count": 10
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-09T17:11:49.234742Z",
     "start_time": "2025-11-09T17:11:49.178935Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Test Model\n",
    "seq_list = [\n",
    "    \"<AGCTTAGCAAGTCCGATC>\",\n",
    "    \"<TTTCCCGGGAAA>\",\n",
    "    \"<ACGTACGTACGT>\"\n",
    "]\n",
    "\n",
    "preds = pipeline.predict(seq_list)\n",
    "for s, p in zip(seq_list, preds):\n",
    "    print(s, \"→\", p)\n"
   ],
   "id": "7daec8bb07b2d129",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<AGCTTAGCAAGTCCGATC> → ncRNA\n",
      "<TTTCCCGGGAAA> → ncRNA\n",
      "<ACGTACGTACGT> → BIOLOGICAL_REGION\n"
     ]
    }
   ],
   "execution_count": 11
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
